frontends/p4/typeChecking/typeUnification.cpp

```diff
@@ -334,23 +374,52 @@ bool TypeUnification::unify(const EqualityConstraint* constraint) {
                     "Cannot unify '%1%' with '%2%'", st->name, strct->name);
             // There is another case, in which each field of the source is unifiable with the
             // corresponding field of the destination, e.g., a struct containing tuples.
-            if (strct->fields.size() != st->fields.size())
+            bool defaultInitializer = false;
+            if (auto unkn = src->to<IR::Type_UnknownStruct>()) {
+                defaultInitializer = unkn->defaultInitializer;
+            }
+            if ((strct->fields.size() != st->fields.size()) && !defaultInitializer) {
                 return constraint->reportError(
                     constraints->getCurrentSubstitution(),
                     "Number of fields %1% in initializer different "
                     "than number of fields in structure %2%: %3% to %4%",
-                    st->fields.size(), strct->fields.size(), st, strct);
-
-            for (const IR::StructField* f : strct->fields) {
-                auto stField = st->getField(f->name);
-                if (stField == nullptr)
-                    return constraint->reportError(constraints->getCurrentSubstitution(),
-                                                   "No initializer for field %1%", f);
-                auto c = constraint->create(f->type, stField->type);
-                c->setError(
-                    "Type of initializer '%1%' does not match type '%2%' of field '%3%' in '%4%'",
-                    { stField->type, f->type, f, strct });
-                constraints->add(c);
+                    st->fields.size(),
+                    strct->fields.size(), st, strct);
+            }
+            if (!defaultInitializer) {
+                for (const IR::StructField* f : strct->fields) {
+                    auto stField = st->getField(f->name);
+                    if (stField == nullptr) {
+                        return constraint->reportError(constraints->getCurrentSubstitution(),
```

Comment:
this code is not about initialization, it is about matching types. the errors should be about not being able to unify two types.